### DISCUSSION TITLE
How to go from a custom module to output html

### DISCUSSION THREAD

**Discussion by RapidOwl** [statiqdev/Discussions#14] [Q&A]
I've read the docs about pipelines and modules, but I can't see how I go from a DTO returned by my custom `ReadDataModule` (gets data from Postgres via the Marten lib). Is there a guide somewhere that shows the interface between what a module outputs (can be any object?) and something that makes it "outputtable"?

In terms of wiring this stuff up, I'd expect modules to wrap their output somehow so I could see what structure the data needs in order to be consumed by `WithOutputWriteFiles`.

**Comment by RapidOwl** [comment]
I'm still utterly baffled by this. Can anyone give me an example? I tried creating a second `Module` to see if I could see what my `ReadDataModule` had put in the execution context (if anything), but I couldn't work out what functions to override in order to get the the execution context.

**Comment by daveaglick** [comment]
Sorry, somehow missed the initial post!

I think the key is understanding that a `IDocument` is essentially a more powerful `IReadOnlyDictionary<string, object>`. There are utility methods like `.GetString(string key)` and `GetInt(string key)`, etc. that help with casting the values, and it contains "content" as well, but from a 10,000 foot view it's basically a glorified dictionary.

When you write a module (or configure an existing module with something like `Config.FromDocument(doc => ...)` you can access that dictionary. Each module generally takes a collection of documents passed in as `IExecutionContext.Inputs` and returns a collection of documents to pass to the next module. Sometimes the return documents are the same as the input documents (for example, if the module is doing something that doesn't modify them), sometimes it's clones of the documents with additional metadata values, and sometimes as with the `ReadDataModule` it's new documents.

The key to understanding how a metadata-heavy document like the outputs from `ReadDataModule` to a file on disk is to understand the relationship between metadata (or the dictionary part of the document) and content. When the `ReadDataModule` outputs documents, it turns each key-value pair of the data (in your case, from a Marten query) into a document with those same key-value pairs. However, you can define a special key using the `.WithContentKey(string contentKey)` method in your custom module constructor to say "this particular key-value should use the value as the content of the document". That content can then be processed by later modules. For example, it could contain a Razor or Markdown template (which would need to be processed by the `RenderRazor` or `RenderMarkdown` modules). Or you could do some further custom processing on it using other custom modules later in the pipeline. But when the document finally gets to the `WriteFiles` module that's added by the `WithOutputWriteFiles` bootstrapper call, it's what's in the document _content_ that gets written to disk.

Some more direct answers in case that helps...

> Is there a guide somewhere that shows the interface between what a module outputs (can be any object?) and something that makes it "outputtable"?

A module is always going to return instance(s) of `IDocument`. That's the encapsulation for all data and content. The `ReadDataModule` base class is nice because it abstracts that from you and instead creates those `IDocument` instances for you from your DTO, but inside every module there's an `Execute...` method that does the heavy lifting. Here's the one from `ReadDataModule` that creates and returns a set of `IDocument`:

![2020-09-21_15h51_59](https://user-images.githubusercontent.com/1020407/93814474-93fadb80-fc22-11ea-93c4-24ab9b1645c3.png)

> In terms of wiring this stuff up, I'd expect modules to wrap their output somehow

You're so close! That's exactly what `IDocument` is for - a wrapper around dictionary-style data with (optional) content.

> so I could see what structure the data needs in order to be consumed by `WithOutputWriteFiles`

The `WithOutputWriteFiles` call adds a `WriteFiles` module to the pipeline. The `WriteFiles` module takes whatever's in the _content_ of the document and writes that. So the missing link here is getting something into the content of the document that represents each of your DTOs that the `WriteFiles` module can write. You can do that with the `WithContentKey` method built into the `ReadDataModule` or using some other means like the `SetContent` module.

> Can anyone give me an example?

I'd be happy to work up a quick example for you - given a DTO with key-value pairs, what kind of output are you trying to produce? A page-per-DTO with an HTML table of the values or something more complicated?

**Comment by RapidOwl** [comment]
Thanks so much for the detailed answer! I'll look into this some more when I get some time.
